File: app/src/main/java/eu/pkgsoftware/babybuddywidgets/BabyPagerAdapter.java
package eu.pkgsoftware.babybuddywidgets;

import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import eu.pkgsoftware.babybuddywidgets.databinding.BabyManagerBinding;
import eu.pkgsoftware.babybuddywidgets.networking.BabyBuddyClient;
import eu.pkgsoftware.babybuddywidgets.networking.ChildrenStateTracker;

// I need to extract this one!!!
class BabyPagerAdapter extends RecyclerView.Adapter<BabyLayoutHolder> {
    private List<BabyLayoutHolder> holders = new ArrayList<>();
    private BabyLayoutHolder activeHolder = null;

    private BaseFragment fragment = null;
    private BabyBuddyClient.Child[] children = null;
    private ChildrenStateTracker stateTracker = null;

    public void postInit(
        BaseFragment fragment,
        BabyBuddyClient.Child[] children,
        ChildrenStateTracker stateTracker
    ) {
        this.fragment = fragment;
        this.stateTracker = stateTracker;
        updateChildren(children);
    }

    public void updateChildren(BabyBuddyClient.Child[] children) {
        this.children = children;
        notifyDataSetChanged();
    }

    @Override
    public @NonNull
    BabyLayoutHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        BabyManagerBinding babyBinding = BabyManagerBinding.inflate(
            fragment.getLayoutInflater(), null, false
        );
        View v = babyBinding.getRoot();
        v.setLayoutParams(new RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT
        ));

        BabyLayoutHolder holder = new BabyLayoutHolder(fragment, babyBinding);
        holders.add(holder);
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull BabyLayoutHolder holder, int position) {
        holder.updateChild(children[position], stateTracker);

        int childIndex = LoggedInFragment.childIndexBySlug(
            children,
            fragment.getMainActivity().getCredStore().getSelectedChild()
        );
        if (childIndex >= 0) {
            if (Objects.equals(children[position], children[childIndex])) {
                activeViewChanged(children[position]);
            }
        }
    }

    @Override
    public void onViewRecycled(@NonNull BabyLayoutHolder holder) {
        holder.clear();
    }

    @Override
    public int getItemCount() {
        if (children == null) {
            return 0;
        }
        return children.length;
    }

    public void activeViewChanged(BabyBuddyClient.Child c) {
        activeHolder = null;
        for (BabyLayoutHolder h : holders) {
            if (Objects.equals(c, h.getChild())) {
                h.updateChild(c, stateTracker);
                activeHolder = h;
            } else {
                h.onViewDeselected();
            }
        }
    }

    public BabyLayoutHolder getActive() {
        return activeHolder;
    }

    public void close() {
        for (BabyLayoutHolder h : holders) {
            h.close();
        }
    }
}

File: app/src/main/java/eu/pkgsoftware/babybuddywidgets/BabyPagerAdapter.kt
package eu.pkgsoftware.babybuddywidgets

import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import androidx.recyclerview.widget.RecyclerView
import eu.pkgsoftware.babybuddywidgets.databinding.BabyManagerBinding
import eu.pkgsoftware.babybuddywidgets.logic.ChildrenStateTracker
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.Child

class BabyPagerAdapter(val stateTracker: ChildrenStateTracker) : RecyclerView.Adapter<BabyLayoutHolder>() {
    var active: BabyLayoutHolder? = null

    private val holders: MutableList<BabyLayoutHolder> = ArrayList()
    private var fragment: BaseFragment? = null
    private var children = emptyArray<Child>()

    init {
        stateTracker.addChildListener({ newChildren ->
            children = newChildren
            notifyDataSetChanged()
        })
    }

    fun postInit(
        fragment: BaseFragment?
    ) {
        this.fragment = fragment
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BabyLayoutHolder {
        val babyBinding = BabyManagerBinding.inflate(
            fragment!!.getLayoutInflater(), null, false
        )
        val v: View = babyBinding.getRoot()
        v.setLayoutParams(
            RelativeLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT
            )
        )

        val holder = BabyLayoutHolder(fragment!!, babyBinding)
        holders.add(holder)
        return holder
    }

    override fun onBindViewHolder(holder: BabyLayoutHolder, position: Int) {
        holder.updateChild(children[position])
    }

    override fun onViewRecycled(holder: BabyLayoutHolder) {
        holder.clear()
    }

    override fun getItemCount(): Int {
        return children.size
    }

    fun activeViewChanged(c: Child) {
        this.active = null
        for (h in holders) {
            if (c == h.child) {
                h.updateChild(c)
                this.active = h
            } else {
                h.onViewDeselected()
            }
        }
    }

    fun close() {
        for (h in holders) {
            h.close()
        }
    }
}
